VitruvioHost/Plugins/Vitruvio/Source/Vitruvio/Private/UnrealCallbacks.cpp

```diff
@@ -150,22 +150,29 @@ void UnrealCallbacks::addMesh(const wchar_t* name, int32_t prototypeId, const wc
 
 	size_t PolygonGroupStartIndex = 0;
 	TArray<Vitruvio::FMaterialAttributeContainer> MeshMaterials;
+	TMap<Vitruvio::FMaterialAttributeContainer, FPolygonGroupID> MeshMaterialMap;
 	for (size_t PolygonGroupIndex = 0; PolygonGroupIndex < faceRangesSize; ++PolygonGroupIndex)
 	{
 		const size_t PolygonFaceCount = faceRanges[PolygonGroupIndex];
 
-		const FPolygonGroupID PolygonGroupId = Description.CreatePolygonGroup();
-
 		Vitruvio::FMaterialAttributeContainer MaterialContainer(materials[PolygonGroupIndex]);
-		const FName MaterialSlot = FName(MaterialContainer.Name);
-		Attributes.GetPolygonGroupMaterialSlotNames()[PolygonGroupId] = MaterialSlot;
 		TMap<FString, double> AvailableUvSetAttributeMap = CreateAvailableUVSetMaterialParameterMap(uvCounts, uvSets);
 		for (auto& AvailableUvSetAttribute : AvailableUvSetAttributeMap)
 		{
 			MaterialContainer.ScalarProperties.Add(AvailableUvSetAttribute);
 		}
-		MeshMaterials.Add(MaterialContainer);
 
+		FPolygonGroupID PolygonGroupId;
+		if (MeshMaterialMap.Contains(MaterialContainer))
+		{
+			PolygonGroupId = MeshMaterialMap[MaterialContainer];
+		}
+		else
+		{
+			MeshMaterials.Add(MaterialContainer);
+			PolygonGroupId = Description.CreatePolygonGroup();
+		}
+		
```

Comment:
The AddMesh function starts to get quite verbose, maybe it would make sense to start using more helper functions like "getPolygonGroupFromMaterial".

Reply:
Good point. I also tried that but it would have created helper methods with a lot of parameters which did not make it more readable. I'll leave this for now.